Fixed e2e tests for HA master.

jszczepkowski opened the pull request:
Set of fixes that allows HA master e2e tests to pass for removal/addition master replicas.

The summary of changes:
- fixed host name in etcd certs,
- added cluster validation after kube-down,
- fixed the number of master replicas in cluster validation,
- made MULTIZONE=true required for HA master deployments, ensured we correctly handle MULTIZONE=true when user wants to create HA master but not kubelets in multiple zones,
- extended verification of master replicas in HA master e2e tests.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36999)
<!-- Reviewable:end -->

jszczepkowski committed:
Fixed e2e tests for HA master.

Set of fixes that allows HA master e2e tests to pass for removal/addition of master replicas.

fgrzadkowski commented:
Can you please explain what you are changing in this PR. It's hard to read it without context.

dims commented:
@k8s-bot gci gke e2e test this

jszczepkowski commented:
Here is the summary of changes:
- fixed host name in etcd certs,
- added restarting master kubelet when the last replica remains after kube-down,
- added cluster validation after kube-down,
- fixed the number of master replicas in cluster validation,
- made MULTIZONE=true required for HA master deployments, ensured we correctly handle MULTIZONE=true when user wants to create HA master but not kubelets in multiple zones,
- extended verification of master replicas in HA master e2e tests.

fgrzadkowski commented on cluster/gce/util.sh:
```diff
@@ -1395,6 +1414,10 @@ function kube-down() {
         --region "${REGION}" \
         --quiet \
         "${MASTER_NAME}"
+      # We need to restart master's kubelet as it uses the external IP which destination has just been changed.
```
As discussed offline we should verify:
* why exactly is it needed
* why we don't need to do it for all other nodes
* what other options do we have? restarting kubelet is pretty intrusive

fgrzadkowski commented on cluster/gce/util.sh:
```diff
@@ -1786,7 +1837,7 @@ function test-setup() {
   detect-project
 
-  if [[ ${MULTIZONE:-} == "true" ]]; then
+  if [[ ${MULTIZONE:-} == "true" && -n ${E2E_ZONES:-} ]]; then
```
Why are you adding it?

jszczepkowski replied:
We need to distinguish two cases: 
1. we want to set-up multimaster cluster with kubelets in a single zone (MULTIZONE=true, E2E_ZONES=false) a
2. we want to set-up cluster with kubelets in many zones (MULTIZONE=ture, E2E_ZONES={list of zones for kubelets})

fgrzadkowski commented on test/e2e/framework/util.go:
```diff
@@ -4011,6 +4011,33 @@ func WaitForClusterSize(c clientset.Interface, size int, timeout time.Duration)
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(20 * time.Second) {
+		nodes, err := c.Core().Nodes().List(api.ListOptions{FieldSelector: fields.Set{
+			"spec.unschedulable": "true",
+		}.AsSelector()})
```
This might list some nodes which are not masters, e.g. those which are drained.

jszczepkowski replied:
Added additional check for prefix.

fgrzadkowski commented on test/e2e/ha_master.go:
```diff
@@ -56,14 +57,21 @@ func verifyRCs(c clientset.Interface, ns string, names []string) {
 func createNewRC(c clientset.Interface, ns string, name string) {
-	_, err := newRCByName(c, ns, name, 1, nil)
+	var err error
+	timeout := 5 * time.Minute
```
Why is it needed?

jszczepkowski replied:
Not needed anymore, removed.

jszczepkowski commented:
@fgrzadkowski 
Comments applied, PTAL

fgrzadkowski commented on test/e2e/framework/util.go:
```diff
@@ -4011,6 +4011,39 @@ func WaitForClusterSize(c clientset.Interface, size int, timeout time.Duration)
 }
 
+// waitForMasters waits until the cluster has the desired number of ready masters in it.
+func WaitForMasters(masterPrefix string, c clientset.Interface, size int, timeout time.Duration) error {
```
How about passing a full regexp for master name and remove checking for unschedulable field?

jszczepkowski replied:
Fixed

fgrzadkowski commented:
Just one last small comment. In general lgtm.

dims commented:
@k8s-bot node e2e test this

jszczepkowski commented:
Part of https://github.com/kubernetes/features/issues/48
